addons/sale_mrp/models/sale_mrp.py

```diff
@@ -35,14 +35,16 @@ def _compute_qty_delivered(self):
                         (b.product_tmpl_id == order_line.product_id.product_tmpl_id and not b.product_id)))
                 if relevant_bom:
                     # In case of dropship, we use a 'all or nothing' policy since 'bom_line_id' was
-                    # not written on a move coming from a PO.
+                    # not written on a move coming from a PO: all moves must be done and the last
+                    # one must be a delivery to the customer (to avoid the return case)
                     # FIXME: if the components of a kit have different suppliers, multiple PO
                     # are generated. If one PO is confirmed and all the others are in draft, receiving
                     # the products for this PO will set the qty_delivered. We might need to check the
                     # state of all PO as well... but sale_mrp doesn't depend on purchase.
                     if dropship:
                         moves = order_line.move_ids.filtered(lambda m: m.state != 'cancel')
-                        if moves and all([m.state == 'done' for m in moves]):
+                        last_done = next((move for move in moves.sorted(key='id', reverse=True) if move.state == 'done'), moves[:0])
```

Comment:
This is still going to fail with backorders if the order with qty > 2 is done with the following flow:

1. Transfer/validate one qty + create backorder (bo1)
2. Return + validate one qty
3.  Validate backorder (bo1) w/ one qty + create another backorder (bo2)
4.  Validate final backorder (bo2)

Final qty will be all are delivered since the last move (bo2) is to the customer even though it should be none are delivered if we continue following the "All or nothing" approach. I'm dubious about using a "look at the last done move" technique because it seems like it will always have edge cases with backorders, but maybe these edge cases are rarer than the case you're trying to fix and it's better to merge this. What do you think? :thinking: